### PR TITLE
Core/AI: enable SmartScripts on Scenes + add scene related actions/events

### LINKED ISSUE #19189: Play a scene from SAI
[//]: # (***************************************************************)
[//]: # (** DON'T DELETE THIS TEMPLATE ELSE YOUR ISSUE WILL BE CLOSED **)
[//]: # (***************************************************************)

**Description:** I think this function have to be integrated. We should be able to play a scene from SAI by using sceneid
We can use it for example in this quest: http://www.wowhead.com/quest=38813/orders-for-your-captains

Looking into sniff when you talk to a Coilskar this should happen:
After selecting Gossip Option -> do Phase Shift -> play scene 1094 -> give kill credit.

**Current behaviour:** (Tell us what happens.) This function is not integrated

**Expected behaviour:** (Tell us what should happen instead.) Function to be integrated

**Steps to reproduce the problem:**

1. 
2. 
3. 

**Branch(es):** master

**TC rev. hash/commit:**  https://github.com/TrinityCore/TrinityCore/commit/bfe47331957b751b1b5cac37cef54364d361d581

**TDB version:**  703.00 + updates

**Operating system:**  Ubuntu server 16.10

[//]: # (This template is for problem reports. For other types of report, edit it accordingly.)
[//]: # (If this is a crash report, include the crashlog with https://gist.github.com/)
[//]: # (For fixes containing C++, create a Pull Request.)

### PR DESCRIPTION
**Pull request by Traesh**
**Changes proposed:**

- Allow SmartScript targeting scenes, with entryOrGuid = sceneId and source_type = 10
- Add new "event_param_string" to smart_script database (usefull by example in SMART_EVENT_SCENE_TRIGGER)

- Add four new SMART_EVENTs :
    - SMART_EVENT_SCENE_START
    - SMART_EVENT_SCENE_TRIGGER
    - SMART_EVENT_SCENE_CANCEL
    - SMART_EVENT_SCENE_COMPLETE

- Add two new SMART_ACTIONs :
    - SMART_ACTION_SCENE_PLAY
    - SMART_ACTION_SCENE_CANCEL

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:** Closes #19189

**Tests performed:** (Does it build, tested in-game, etc.)
Built, tested ingame with following SQL, should spawn creature 1 on scene 1181 triggering event (.cast spell 201179)
```sql
INSERT INTO `smart_scripts` VALUES (1181,10,0,0,79,0,100,0,0,0,0,0,'BURN PLAYER',12,1,1,0,0,0,0,1,0,0,0,0,0,0,0,'');

```

**Known issues and TODO list:** (add/remove lines as needed)

- [ ]

### CODE REVIEW

File: src/server/game/AI/SmartScripts/SmartScriptMgr.h [hunk 1]
```diff
@@ -558,8 +558,19 @@ enum SMART_ACTION
     SMART_ACTION_RANDOM_SOUND                       = 115,    // soundId1, soundId2, soundId3, soundId4, soundId5, onlySelf
     SMART_ACTION_SET_CORPSE_DELAY                   = 116,    // timer
     SMART_ACTION_DISABLE_EVADE                      = 117,    // 0/1 (1 = disabled, 0 = enabled)
-
-    SMART_ACTION_END                                = 118
+    SMART_ACTION_SET_CAN_FLY                        = 119,    // 0/1 (3.3.5 only, NYI)
```
Comment (Aokromes) [inline src/server/game/AI/SmartScripts/SmartScriptMgr.h]: 118 to 127 are on used on 3.3.5a.

Comment (Traesh) [inline src/server/game/AI/SmartScripts/SmartScriptMgr.h]: Yup, already saw it in another PR, that's why i set placeholder with comments (3.3.5 only, NYI) :)

Maybe a bad idea, will remove it if required

**Comment by Shauren** [comment]
can i just say plsno to player sai?

**Comment by Traesh** [comment]
Yup, can i ask why ? :p

i feel like this could be usefull in some case

**Comment by LuigiElleBalotta** [comment]
Yes that could be useful

**Comment by streetrat** [comment]
this is a little overkill
scene scripting will be needed, but using sai vs players is just bad
problems:
- you cant use any timed stuff, only if you hook into player::update, just like for creatures
- worst: ALL player scripts will be loaded to ALL players... imagine a year later, when all scenes are scripted, and you have 2k sai scripts for those scenes.. you have 1k player with 2k sai.. and probably 99.9999% nothing will be exacuted

if you want to use sai + players + events, then instead of hacking into players, create a new source type (scenes) and use the same hooks as you did, the invoker will be the player, this is how sai should be used

**Comment by Traesh** [comment]
@streetrat You got me at "imagine a year later, when all scenes are scripted", you now have one year to script them all :p

I agree on overkill for scenes, i'll switch behavior to a more conventional way

**Comment by tkrokli** [comment]
Note to the dev merging this issue in the future: `Closes #19189` is missing in the commit message. This can be added in the github merge dialog (the final commit message needs editing anyway).

**Comment by Traesh** [comment]
@Shauren It should be better now, player sai removed :)

**Pull request closed without merging**